client/src/legacy/LeftAndMain.js

```diff
@@ -460,77 +460,93 @@ $.entwine('ss', function($) {
       // default to first button if none given - simulates browser behaviour
       if(!button) button = this.find('.btn-toolbar :submit:first');
 
-      form.trigger('beforesubmitform');
-      this.trigger('submitform', {form: form, button: button});
+      var beforeSubmitFormEventData = {
+        // array of promises that must resolve(true) before the form is submitted
+        promises: [],
+        // callbacks that are called on ajax success after submitted the form
+        onAjaxSuccessCallbacks: [],
+      };
+      form.trigger('beforesubmitform', beforeSubmitFormEventData);
 
-      // set button to "submitting" state
-      $(button).addClass('btn--loading loading');
-      $(button).prop('disabled', true);
+      Promise.all(beforeSubmitFormEventData.promises).then(function(results) {
 
-      if($(button).is('button')) {
-        $(button).data('original-text', $(button).text());
+        for (const result of results) {
```

Comment:
This block is newly added - if any of the promises resolve(false) then the form will not submit